cmake/Modules/FindX11.cmake

```diff
@@ -0,0 +1,898 @@
+      INTERFACE_LINK_LIBRARIES "X11::Xext;X11::X11")
+  endif ()
+
+  if (X11_Xt_FOUND AND NOT TARGET X11::Xt)
+    add_library(X11::Xt UNKNOWN IMPORTED)
+    set_target_properties(X11::Xt PROPERTIES
+      IMPORTED_LOCATION "${X11_Xt_LIB}"
+      INTERFACE_INCLUDE_DIRECTORIES "${X11_Xt_INCLUDE_PATH}"
+      INTERFACE_LINK_LIBRARIES "X11::ICE;X11::SM;X11::X11")
+  endif ()
+
+  if (X11_Xutil_FOUND AND NOT TARGET X11::Xutil)
+    add_library(X11::Xutil INTERFACE IMPORTED)
+    set_target_properties(X11::Xutil PROPERTIES
+      INTERFACE_INCLUDE_DIRECTORIES "${X11_Xutil_INCLUDE_PATH}"
+      # libX11 contains the implementations for functions in the Xutil.h
+      # header.
+      INTERFACE_LINK_LIBRARIES "X11::X11")
+  endif ()
+
+  if (X11_Xv_FOUND AND NOT TARGET X11::Xv)
+    add_library(X11::Xv UNKNOWN IMPORTED)
+    set_target_properties(X11::Xv PROPERTIES
+      IMPORTED_LOCATION "${X11_Xv_LIB}"
+      INTERFACE_INCLUDE_DIRECTORIES "${X11_Xv_INCLUDE_PATH}"
+      INTERFACE_LINK_LIBRARIES "X11::Xext;X11::X11")
+  endif ()
+
+  mark_as_advanced(
+    X11_X11_INCLUDE_PATH
+    X11_X11_LIB
+    X11_Xext_INCLUDE_PATH
+    X11_Xext_LIB
+    X11_Xau_LIB
+    X11_Xau_INCLUDE_PATH
+    X11_xcb_LIB
+    X11_xcb_INCLUDE_PATH
+    X11_xcb_xkb_LIB
+    X11_X11_xcb_LIB
+    X11_X11_xcb_INCLUDE_PATH
+    X11_Xlib_INCLUDE_PATH
+    X11_Xutil_INCLUDE_PATH
+    X11_Xcomposite_INCLUDE_PATH
+    X11_Xcomposite_LIB
+    X11_Xfixes_LIB
+    X11_Xfixes_INCLUDE_PATH
+    X11_Xrandr_LIB
+    X11_Xrandr_INCLUDE_PATH
+    X11_Xdamage_LIB
+    X11_Xdamage_INCLUDE_PATH
+    X11_Xrender_LIB
+    X11_Xrender_INCLUDE_PATH
+    X11_XRes_LIB
+    X11_XRes_INCLUDE_PATH
+    X11_Xxf86misc_LIB
+    X11_Xxf86misc_INCLUDE_PATH
+    X11_Xxf86vm_LIB
+    X11_Xxf86vm_INCLUDE_PATH
+    X11_Xi_LIB
+    X11_Xi_INCLUDE_PATH
+    X11_Xinerama_LIB
+    X11_Xinerama_INCLUDE_PATH
+    X11_Xtst_LIB
+    X11_Xtst_INCLUDE_PATH
+    X11_Xcursor_LIB
+    X11_Xcursor_INCLUDE_PATH
+    X11_dpms_INCLUDE_PATH
+    X11_Xt_LIB
+    X11_Xt_INCLUDE_PATH
+    X11_Xdmcp_LIB
+    X11_LIBRARIES
+    X11_Xaccessrules_INCLUDE_PATH
+    X11_Xaccessstr_INCLUDE_PATH
+    X11_Xdmcp_INCLUDE_PATH
+    X11_Xkb_INCLUDE_PATH
+    X11_Xkblib_INCLUDE_PATH
+    X11_xkbcommon_INCLUDE_PATH
+    X11_xkbcommon_LIB
+    X11_xkbcommon_X11_INCLUDE_PATH
+    X11_xkbcommon_X11_LIB
+    X11_xkbfile_INCLUDE_PATH
+    X11_xkbfile_LIB
+    X11_Xmu_INCLUDE_PATH
+    X11_Xmu_LIB
+    X11_Xss_INCLUDE_PATH
+    X11_Xss_LIB
+    X11_Xpm_INCLUDE_PATH
+    X11_Xpm_LIB
+    X11_Xft_LIB
+    X11_Xft_INCLUDE_PATH
+    X11_Xshape_INCLUDE_PATH
+    X11_Xv_LIB
+    X11_Xv_INCLUDE_PATH
+    X11_XShm_INCLUDE_PATH
+    X11_ICE_LIB
+    X11_ICE_INCLUDE_PATH
+    X11_SM_LIB
+    X11_SM_INCLUDE_PATH
+    X11_XSync_INCLUDE_PATH
+    X11_Xaw_LIB
+    X11_Xaw_INCLUDE_PATH
+  )
+  set(CMAKE_FIND_FRAMEWORK ${CMAKE_FIND_FRAMEWORK_SAVE})
+  set(CMAKE_REQUIRED_QUIET ${CMAKE_REQUIRED_QUIET_SAVE})
+endif ()
```

Comment:
Could do it, but this is cmake’s official FindX11 anyway (to help Linux on 3.16 which doesn’t have the same targets).

Reply:
Ugh.  That's the only one I noticed, but their entire FindX11.cmake module mixes `endif()` and `endif ()` throughout.  Gross.  =/

Reply:
Will probably be changed by cmake-format anyway I guess, I’ll watch this one before pushing that commit.

Reply:
Is there a reason we need to include it if CMake includes it?  Did we need a specific version?

Reply:
Yep the finder has been updated between 3.16 and more recent versions to supply targets like `X11::X11_xcb` which many Linux targets depend on. In the end it was easier to ‘backport’ the more modern finder than add lots of custom code myself (and fiddle around with custom link and include variables).

Reply:
Do you happen to know in what version of CMake it satisfies your requirements?  I'd like to know at which point we could remove it.

Reply:
Will have to check their repo first. It should be "fine" in 3.19 already.

Reply:
If we recommend installing a newer cmake via Kitware's apt repo (which I thought we'd discussed anyway), then we don't need to worry about this.

If we're relying on the defaults that come with Ubuntu, then Ubuntu 20.04 still has CMake 3.16.x (https://launchpad.net/ubuntu/focal/+package/cmake).  Ubuntu 21.04 is using CMake 3.18.x by default.  If we need CMake 3.19 for this, the earliest we'd be able to rely on the Ubuntu default here is probably 21.10 or later.

Anyway, what I'm looking for is a distinct note somewhere that reminds us to revisit dropping it in the future.

Reply:
Iirc installing a newer cmake via Kitware repo is not possible when packaging ppa's on Launchpad, so the build system needs to work with the default version available on Ubuntu 20.04.

Reply:
@Gol-D-Ace ?

Reply:
I'm afraid I don't follow - except for `FindX11.cmake` all other finders are for packages/dependencies that CMake doesn't have its own finders for. And this finder is only required until the baseline CMake version for Linux builds provides the same targets as this version (Cmake 3.18 was the first version supplying the `X11::x11_xcb` target).

Reply:
> only required until the baseline CMake version for Linux builds provides the same targets as this version

Which is about 4 years out it seems. During which time if breaking changes are made in any of the Finders distributed with cmakes in the future, like we saw cmake did with FindThreads, things will inexplicably break. 

> I'm afraid I don't follow

So making it a hard requirement to not include finders and instead raise the version (as much as I might want personally) would mean we couldnt package except for ubuntu 21.10 and above (aka, unmergable). So we probably just accept the fact that we hope cmake doesnt change FindX11.cmake or any other bundled finders that depend on X11 in a breaking way. So I personally think it's probably better to have the PR than not, but ill be really sad if cmake breaks this.

Reply:
Got it - the alternative would be to use the "official" X11 finder of Cmake 3.16 and manually supply the `X11_xcb` headers+library in the CMakeLists that require it based on the library and include paths supplied by the main X11 component. Iirc that wouldn't even be that much extra effort (it's just "ugly").

Alternatively this could be patched into the custom XCB finder we use because we link specific additional components that are not exposed by the official X11 finder (even in CMake 3.20).

Reply:
I honestly do not know what this particular fun library does, but it ships with its own pkg-config (x11-xcb) on my system. So we could just add a finder just for it until we can rely on FindX11 from cmake 3.18.

Reply:
I ended up reinstating the "old" `X11_XCB` finder that we ship with already and just added the target definition.